### PR TITLE
Add options to set PR to Approved or Needs Work

### PR DESCRIPTION
**Pull request by mrdima**
Based on the work of https://github.com/nemccarthy/stash-pullrequest-builder-plugin/pull/123
We already use the original PR of tettaji successfully for the last one and a half years. In our workflow we do not automatically merge by only a successful jenkins build, we also require 2 human peer reviews next to the jenkins (build) review. We merge "manually" when we desire to do so after these 3 approvals.

### CODE REVIEW

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashMarkStatus.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        Result result, StashRepository repository) {
+        if(approveOnBuildSuccessful && result == Result.SUCCESS) {
+            repository.markStatus(pullRequestId, "APPROVED");
+        }
+
+        if(needsWorkOnBuildFailure && result == Result.FAILURE) {
+            repository.markStatus(pullRequestId, "NEEDS_WORK");
+        }
+    }
+}
```
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashMarkStatus.java:18]: I don't think `StashMarkStatus` needs to be a separate class. I'd rather stop proliferation of classes that have no state or map 1-to-1 to the existing classes.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java [hunk 2]
```diff
@@ -64,11 +65,13 @@
     private String project;
     private String repositoryName;
     private Credentials credentials;
+    private String username;
 
 
     public StashApiClient(String stashHost, String username, String password, String project, String repositoryName, boolean ignoreSsl) {
         this.credentials = new UsernamePasswordCredentials(username, password);
         this.project = project;
+	this.username = username;
```
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:74]: We are going to reformat the whole codebase and enable automatic formatting, which would make formatting issues irrelevant. However, please use reasonable formatting in your submissions to facilitate reviews.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java [hunk 3]
```diff
@@ -414,6 +428,94 @@ public String call() throws Exception {
...
 
+    private String putRequest(String path, String status) throws UnsupportedEncodingException {
+        logger.log(Level.FINEST, "PR-PUT-REQUEST:" + path + " with: " + status);
+        HttpClient client = getHttpClient();
+	client.getState().setCredentials(AuthScope.ANY, credentials);
+
+        PutMethod httpput = new PutMethod(path);
+        //http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html; section 14.10.
+        //tells the server that we want it to close the connection when it has sent the response.
+        //address large amount of close_wait sockets client and fin sockets server side
```
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:439]: Please use complete sentences in large comments and add space after `//` for readability.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java [hunk 4]
```diff
@@ -414,6 +428,94 @@ public String call() throws Exception {
...
+        }
+
+         String response = "";
+        FutureTask<String> httpTask = null;
+        Thread thread;
+
+         try {
+            //Run the http request in a future task so we have the opportunity
+            //to cancel it if it gets hung up; which is possible if stuck at
+            //socket native layer.  see issue JENKINS-30558
```
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:465]: That issue is closed with "cannot reproduce". https://issues.jenkins-ci.org/browse/JENKINS-30558

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly [hunk 5]
```diff
@@ -52,5 +52,11 @@
     <f:entry title="Phrase to request a (re-)build" field="ciBuildPhrases">
       <f:textbox default="test this please"/>
     </f:entry>
+    <f:entry title="Approve PR on build success" field="approveOnBuildSuccessful">
+      <f:checkbox default="false"/>
+    </f:entry>
+    <f:entry title="Mark PR with Needs Work on build failure" field="needsWorkOnBuildFailure">
+      <f:checkbox default="false"/>
+    </f:entry>
```
Comment (proski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly:60]: Please add html help files for those fields.

**Review by proski — CHANGES_REQUESTED** [review]
Good idea, but the implementation is somewhat dated. The formatting is bad in many places. A class is added for no good reason. Help files are missing for the new options.

**Comment by jakub-bochenski** [comment]
I second @proski  comments.

Please take care to add good help/documentation. It's great that this works for you, but describing the feature well is key to enabling others to make use of it.

**Review by jakub-bochenski — CHANGES_REQUESTED** [review]
I second @proski  comments.

Please take care to add good help/documentation. It's great that this works for you, but describing the feature well is key to enabling others to make use of it.

**Comment by jakub-bochenski** [comment]
With the changes just merged to master local build should format the files automatically

**Comment by dawidmalina** [comment]
@jakub-bochenski, @proski I think that @mrdima just grabbed original pull request https://github.com/nemccarthy/stash-pullrequest-builder-plugin/pull/123 as is without any code modification so he is not directly guilty :)

As I'm also interested in this change I can give a hand on this. Just please guide us through all required by you code change which should be applied.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java (hunk 3 above)
Comment (dawidmalina) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:439]: The same comment is used also in 2 other places. Do you think it should be also fixed there? Btw it just a extract from html rfc about closing connection - IMHO this comment can be removed entirely.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java (hunk 4 above)
Comment (dawidmalina) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:465]: This ticket is mentioned in few other places should we also remove it entirely?

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashMarkStatus.java (hunk 1 above)
Comment (dawidmalina) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashMarkStatus.java:18]: Where is the best place for this handler?

**Comment by dawidmalina** [comment]
I've created new pull request with provided resolved conflict, formatting and help files in https://github.com/jenkinsci/stash-pullrequest-builder-plugin/pull/56

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashMarkStatus.java (hunk 1 above)
Comment (proski) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashMarkStatus.java:18]: I would just put that code where it's being called. If the code gets too long, we care refactor it across more natural lines.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java (hunk 3 above)
Comment (proski) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:439]: Let's remove it then.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java (hunk 4 above)
Comment (proski) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:465]: At least don't add it blindly. We can cleanup the existing comments in a separate PR.

**Comment by proski** [comment]
Let's close this PR in favor of PR #56 to avoid parallel discussions. It's out of date anyway.

**Pull request closed without merging**